billing/models.py

```diff
@@ -1,3 +1,91 @@
+
+from billing.common.constants import BillingConstants
+from billing.common.constants import CurrencyConstants
+from billing.common.strings import AccountingMethodText
+from billing.common.strings import BillingDetailsAbstractModelText
+from billing.common.strings import BillingValidationErrorText
+from managers.models import Project
+from users.common.fields import ChoiceEnum
+from users.models import CustomUser
+from utils.models import BaseModel
+
+
+class BillingDetailsAbstractQuerySet(models.QuerySet):
+    pass
+
+
+class BillingDetailsAbstract(BaseModel):
+    class Meta:
+        abstract = True
+
+    class Currency(ChoiceEnum):
+        PLN = CurrencyConstants.CurrencyDict["PLN"]["Code"]
+        USD = CurrencyConstants.CurrencyDict["USD"]["Code"]
+        EUR = CurrencyConstants.CurrencyDict["EUR"]["Code"]
+
+    class AccountingMethod(ChoiceEnum):
+        HOURLY = AccountingMethodText.HOURLY
+        MONTHLY = AccountingMethodText.MONTHLY
+
+    user = models.ForeignKey(CustomUser, on_delete=models.PROTECT, related_name="%(class)s")
+    value = models.PositiveSmallIntegerField(BillingDetailsAbstractModelText.VALUE, blank=False, null=False)
+    currency = models.CharField(
+        max_length=BillingConstants.CURRENCY_MAX_LENGTH.value, choices=Currency.choices(), default=Currency.PLN
+    )
+    accounting_method = models.CharField(
+        max_length=BillingConstants.ACCOUNTING_METHOD_MAX_LENGTH.value,
+        choices=AccountingMethod.choices(),
+        default=AccountingMethod.HOURLY.name,
+    )
+    from_date = models.DateTimeField(
+        BillingDetailsAbstractModelText.FROM_DATE, auto_now_add=False, blank=False, null=False
+    )
+    to_date = models.DateTimeField(BillingDetailsAbstractModelText.TO_DATE, auto_now_add=False, blank=True, null=True)
+
+    objects = BillingDetailsAbstractQuerySet.as_manager()
+
+    def clean(self) -> None:
+        super().clean()
+
+        if self.to_date is not None and self.to_date <= self.from_date:
+            raise ValidationError(message=BillingValidationErrorText.VALIDATION_ERROR_FROM_DATE_LATER_THAN_TO_DATE)
+        if self.to_date is None and not self.user.is_active:
+            raise ValidationError(message=BillingValidationErrorText.VALIDATION_ERROR_TO_DATE_EMPTY_FOR_INACTIVE_USER)
+
+    @property
+    def get_currency_name(self) -> str:
+        return CurrencyConstants.CurrencyDict[self.currency]["Name"]
+
+    @property
+    def get_currency_code(self) -> str:
+        return CurrencyConstants.CurrencyDict[self.currency]["Code"]
+
+    @property
+    def get_currency_symbol(self) -> str:
+        return CurrencyConstants.CurrencyDict[self.currency]["Symbol"]
+
+
+class UserSalaryDetails(BillingDetailsAbstract):
+    def save(self, *args: Any, **kwargs: Any) -> None:
+        self._fix_user_has_two_ongoing_salaries()
+        super(UserSalaryDetails, self).save(*args, **kwargs)
+
+    def _fix_user_has_two_ongoing_salaries(self) -> None:
```

Comment:
It's better to keep such things in clean then in save :)

Please also change name of it because `fix` is not a perfect word for such things. It'd be better to have something like `check_ongoing_salaries` or something like that :)

Reply:
Ok